In [9]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

df = pd.read_csv("../data/survey.csv")

df.head()

Timestamp  Age  Gender         Country state self_employed  \
0  2014-08-27 11:29:31   37  Female   United States    IL           NaN   
1  2014-08-27 11:29:37   44       M   United States    IN           NaN   
2  2014-08-27 11:29:44   32    Male          Canada   NaN           NaN   
3  2014-08-27 11:29:46   31    Male  United Kingdom   NaN           NaN   
4  2014-08-27 11:30:22   31    Male   United States    TX           NaN   

  family_history treatment work_interfere    no_employees  ...  \
0             No       Yes          Often            6-25  ...   
1             No        No         Rarely  More than 1000  ...   
2             No        No         Rarely            6-25  ...   
3            Yes       Yes          Often          26-100  ...   
4             No        No          Never         100-500  ...   

                leave mental_health_consequence phys_health_consequence  \
0       Somewhat easy                        No                      No   
1          Don't know                     Maybe                      No   
2  Somewhat difficult                        No                      No   
3  Somewhat difficult                       Yes                     Yes   
4          Don't know                        No                      No   

      coworkers supervisor mental_health_interview phys_health_interview  \
0  Some of them        Yes                      No                 Maybe   
1            No         No                      No                    No   
2           Yes        Yes                     Yes                   Yes   
3  Some of them         No                   Maybe                 Maybe   
4  Some of them        Yes                     Yes                   Yes   

  mental_vs_physical obs_consequence comments  
0                Yes              No      NaN  
1         Don't know              No      NaN  
2                 No              No      NaN  
3                 No             Yes      NaN  
4         Don't know              No      NaN  

[5 rows x 27 columns]

In [10]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Timestamp                  1259 non-null   object
 1   Age                        1259 non-null   int64 
 2   Gender                     1259 non-null   object
 3   Country                    1259 non-null   object
 4   state                      744 non-null    object
 5   self_employed              1241 non-null   object
 6   family_history             1259 non-null   object
 7   treatment                  1259 non-null   object
 8   work_interfere             995 non-null    object
 9   no_employees               1259 non-null   object
 10  remote_work                1259 non-null   object
 11  tech_company               1259 non-null   object
 12  benefits                   1259 non-null   object
 13  care_options               1259 non-null   object
 14  wellness

Age
count  1.259000e+03
mean   7.942815e+07
std    2.818299e+09
min   -1.726000e+03
25%    2.700000e+01
50%    3.100000e+01
75%    3.600000e+01
max    1.000000e+11

#### Here we can clearly see that age column from the dataset is corrupted with incorrect data, see the mean is around 7.94 x 10^7 which is obviously impossible for a human, also standard deviation which is the spread of data on mean is also very high. Minimum age is below negative which is not possible and maximum age is 10^11 which is out of ths world.

## Cleaning data

In [14]:
df.isnull().sum().sort_values(ascending=False)

comments                     1095
state                         515
work_interfere                264
self_employed                  18
Gender                          0
Timestamp                       0
Age                             0
family_history                  0
treatment                       0
no_employees                    0
Country                         0
remote_work                     0
tech_company                    0
care_options                    0
benefits                        0
seek_help                       0
anonymity                       0
leave                           0
wellness_program                0
mental_health_consequence       0
phys_health_consequence         0
supervisor                      0
coworkers                       0
mental_health_interview         0
phys_health_interview           0
mental_vs_physical              0
obs_consequence                 0
dtype: int64

In [19]:
# Droppin columns with too many missing values
df.drop(columns=['state'], axis=1, inplace=True)

In [23]:
df['work_interfere'].fillna('Don’t know', inplace=True)

df['self_employed'].fillna('NAN', inplace=True)

In [25]:
df.head()

df.isnull().sum().sort_values(ascending=False)

Timestamp                    0
Age                          0
Gender                       0
Country                      0
self_employed                0
family_history               0
treatment                    0
work_interfere               0
no_employees                 0
remote_work                  0
tech_company                 0
benefits                     0
care_options                 0
wellness_program             0
seek_help                    0
anonymity                    0
leave                        0
mental_health_consequence    0
phys_health_consequence      0
coworkers                    0
supervisor                   0
mental_health_interview      0
phys_health_interview        0
mental_vs_physical           0
obs_consequence              0
dtype: int64

#### Thus, now there is no missing values in the dataset.

### Check for inconsistent columns

In [26]:
for col in df.columns:
  if df[col].dtype == "object":
    print(f"\n Unique Values in '{col}:")
    print(df[col].unique())


 Unique Values in 'Timestamp:
['2014-08-27 11:29:31' '2014-08-27 11:29:37' '2014-08-27 11:29:44' ...
 '2015-11-07 12:36:58' '2015-11-30 21:25:06' '2016-02-01 23:04:31']

 Unique Values in 'Gender:
['Female' 'M' 'Male' 'male' 'female' 'm' 'Male-ish' 'maile' 'Trans-female'
 'Cis Female' 'F' 'something kinda male?' 'Cis Male' 'Woman' 'f' 'Mal'
 'Male (CIS)' 'queer/she/they' 'non-binary' 'Femake' 'woman' 'Make' 'Nah'
 'All' 'Enby' 'fluid' 'Genderqueer' 'Female ' 'Androgyne' 'Agender'
 'cis-female/femme' 'Guy (-ish) ^_^' 'male leaning androgynous' 'Male '
 'Man' 'Trans woman' 'msle' 'Neuter' 'Female (trans)' 'queer'
 'Female (cis)' 'Mail' 'cis male' 'A little about you' 'Malr' 'p' 'femail'
 'Cis Man' 'ostensibly male, unsure what that really means']

 Unique Values in 'Country:
['United States' 'Canada' 'United Kingdom' 'Bulgaria' 'France' 'Portugal'
 'Netherlands' 'Switzerland' 'Poland' 'Australia' 'Germany' 'Russia'
 'Mexico' 'Brazil' 'Slovenia' 'Costa Rica' 'Austria' 'Ireland' 'India'
 

#### From above observations, we can clearly see only gender column contains inconsistent entries and timestamp contains too many unique entries which is also irrelevant to us.

In [28]:
df.drop("Timestamp", axis=1, inplace=True)

### Standardize Gender

In [29]:
def clean_gender(gender):
  gender = str(gender).strip().lower()
  male_terms = ['male', 'm', 'man', 'msle', 'mail', 'malr', 'mal', 'cis male', 'cis man', 'guy', 'make']
  female_terms = ['female', 'f', 'woman', 'femake', 'cis female', 'cis-female/femme', 'female (cis)', 'female (trans)', 'trans-female']

  if gender in male_terms:
    return "Male"
  
  if gender in female_terms:
    return "Female"
  
  else:
    return "Other"

df['gender'] = df['Gender'].apply(clean_gender)

In [30]:
df['gender'].head()

0    Female
1      Male
2      Male
3      Male
4      Male
Name: gender, dtype: object

In [35]:
df.drop(columns=['Gender'], errors='ignore', inplace=True)

In [37]:
df.head()

Age         Country self_employed family_history treatment work_interfere  \
0   37   United States           NAN             No       Yes          Often   
1   44   United States           NAN             No        No         Rarely   
2   32          Canada           NAN             No        No         Rarely   
3   31  United Kingdom           NAN            Yes       Yes          Often   
4   31   United States           NAN             No        No          Never   

     no_employees remote_work tech_company    benefits  ...  \
0            6-25          No          Yes         Yes  ...   
1  More than 1000          No           No  Don't know  ...   
2            6-25          No          Yes          No  ...   
3          26-100          No          Yes          No  ...   
4         100-500         Yes          Yes         Yes  ...   

                leave mental_health_consequence phys_health_consequence  \
0       Somewhat easy                        No                      No   
1          Don't know                     Maybe                      No   
2  Somewhat difficult                        No                      No   
3  Somewhat difficult                       Yes                     Yes   
4          Don't know                        No                      No   

      coworkers supervisor mental_health_interview phys_health_interview  \
0  Some of them        Yes                      No                 Maybe   
1            No         No                      No                    No   
2           Yes        Yes                     Yes                   Yes   
3  Some of them         No                   Maybe                 Maybe   
4  Some of them        Yes                     Yes                   Yes   

  mental_vs_physical obs_consequence  gender  
0                Yes              No  Female  
1         Don't know              No    Male  
2                 No              No    Male  
3                 No             Yes    Male  
4         Don't know              No    Male  

[5 rows x 24 columns]